pkg/sql/parser/aggregate_builtins.go

```diff
@@ -813,6 +857,282 @@ func (a *intervalSumAggregate) Result() (Datum, error) {
+	sqrDiff apd.Decimal
+
+	// Variables used as scratch space within iterations.
+	tmpCount apd.Decimal
+	tmpMean  apd.Decimal
+	delta    apd.Decimal
+	tmp      apd.Decimal
+}
+
+func newDecimalSumSqrDiffsAggregate(_ []Type, _ *EvalContext) AggregateFunc {
+	// Use extra internal precision during squared difference to protect against
+	// order changes that can happen in dist SQL. The additional 3 here should
+	// allow for correctness up to 1000 more worst case inputs than non-worst
+	// case inputs. See #13689 for more analysis and other algorithms.
+	c := DecimalCtx.WithPrecision(DecimalCtx.Precision + 3)
+	ed := apd.MakeErrDecimal(c)
+	return &decimalSumSqrDiffsAggregate{
+		ed: &ed,
+	}
+}
+
+func (a *decimalSumSqrDiffsAggregate) Add(_ context.Context, sqrDiffD Datum, otherArgs ...Datum) error {
+	sumD := otherArgs[0]
+	countD := otherArgs[1]
+	if sqrDiffD == DNull || sumD == DNull || countD == DNull {
+		return nil
+	}
+	sqrDiff := &sqrDiffD.(*DDecimal).Decimal
+	sum := &sumD.(*DDecimal).Decimal
+	a.tmpCount.SetInt64(int64(*countD.(*DInt)))
+
+	a.ed.Quo(&a.tmpMean, sum, &a.tmpCount)
+	a.ed.Sub(&a.delta, &a.tmpMean, &a.mean)
+
+	// Compute the sum of Knuth/Welford sum of squared differences from the mean in a
+	// single pass. Adapted from sum of RunningStats in https://www.johndcook.com/blog/skewness_kurtosis
+	// and our implementation of NumericStats https://github.com/cockroachdb/cockroach/pull/17728.
+
+	// This is logically equivalent to
+	//   sqrDiff + delta * delta * tmpCount * a.count / (tmpCount + a.count)
+	// where the expression is computed from RIGHT to LEFT.
+	a.ed.Add(&a.tmp, &a.tmpCount, &a.count)
+	a.ed.Quo(&a.tmp, &a.count, &a.tmp)
+	a.ed.Mul(&a.tmp, &a.tmpCount, &a.tmp)
+	a.ed.Mul(&a.tmp, &a.delta, &a.tmp)
+	a.ed.Mul(&a.tmp, &a.delta, &a.tmp)
+	a.ed.Add(&a.tmp, sqrDiff, &a.tmp)
+	// Update running squared difference.
+	a.ed.Add(&a.sqrDiff, &a.sqrDiff, &a.tmp)
+
+	// Update total count.
+	a.ed.Add(&a.count, &a.count, &a.tmpCount)
+
+	// This is logically equivalent to
+	//   delta * tmpCount / a.count
+	// where the expression is computed from LEFT to RIGHT.
+	// Note `a.count` is now the total count (includes tmpCount).
+	a.ed.Mul(&a.tmp, &a.delta, &a.tmpCount)
+	a.ed.Quo(&a.tmp, &a.tmp, &a.count)
+	// Update running mean.
+	a.ed.Add(&a.mean, &a.mean, &a.tmp)
+
+	return a.ed.Err()
+}
+
+func (a *decimalSumSqrDiffsAggregate) Result() (Datum, error) {
+	if a.count.Cmp(decimalOne) < 0 {
+		return DNull, nil
+	}
+	dd := &DDecimal{a.sqrDiff}
+	// Remove trailing zeros. Depending on the order in which the input
+	// is processed, some number of trailing zeros could be added to the
+	// output. Remove them so that the results are the same regardless of order.
+	dd.Decimal.Reduce(&dd.Decimal)
+	return dd, nil
+}
+
+// Close is part of the AggregateFunc interface.
+func (a *decimalSumSqrDiffsAggregate) Close(context.Context) {}
+
+// TODO(richardwu): refactor (int|decimal)VarianceAggregate to varianceAggregate that
```

Comment:
There is a lot of copied code here since we needed a way to aggregate only the sum of squared differences in the local stages (as opposed to locally calculated standard deviation/variances).

In a later PR I will consolidate the two implementations of variances (final and non-final) into one that uses the common `sqrDiffAggregate` functions.